### PR TITLE
CSHARP-5311: Remove deprecation from BSON area

### PR DESCRIPTION
**Pull request by papafe**

### CODE REVIEW

File: src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs [hunk 1]
```diff
@@ -33,7 +33,7 @@ public enum BsonBinarySubType
         /// <summary>
         /// Obsolete binary data subtype (use Binary instead).
         /// </summary>
-        [Obsolete("Use Binary instead")]
+        [Obsolete("Use Binary instead")]  //TODO Should we remove this?
```
Comment (papafe) [inline src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs:36]: Not sure if we should remove this here

File: src/MongoDB.Bson/ObjectModel/BsonDocument.cs [hunk 2]
```diff
@@ -167,7 +132,7 @@ public BsonDocument(IEnumerable<BsonElement> elements)
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors")]
         public BsonDocument(params BsonElement[] elements)
```
Comment (papafe) [inline src/MongoDB.Bson/ObjectModel/BsonDocument.cs:133]: Not sure if we want to remove this constructor, as in some cases it's easier than to create an enumerable and pass it. On the other hand maybe we do not want to have so many similar constructor

File: src/MongoDB.Bson/IO/JsonOutputMode.cs [hunk 3]
```diff
@@ -25,7 +25,7 @@ public enum JsonOutputMode
         /// <summary>
         /// Output strict JSON (an obsolete output mode that is similar to canonical extended JSON).
         /// </summary>
-        [Obsolete("Use CanonicalExtendedJson instead.")]
+        [Obsolete("Use CanonicalExtendedJson instead.")]  //TODO Should we remove this?
```
Comment (papafe) [inline src/MongoDB.Bson/IO/JsonOutputMode.cs:28]: Not sure if we should remove this

File: src/MongoDB.Bson/Serialization/Serializers/DictionarySerializerBase.cs [hunk 4]
```diff
@@ -605,7 +605,7 @@ protected virtual ICollection<KeyValuePair<TKey, TValue>> CreateAccumulator()
         /// Creates the instance.
         /// </summary>
         /// <returns>The instance.</returns>
-        [Obsolete("CreateInstance is deprecated. Please use CreateAccumulator instead.")]
+        [Obsolete("CreateInstance is deprecated. Please use CreateAccumulator instead.")]  //TODO Not sure what we want to do with this, it's used by the CreateAccumulator method
```
Comment (papafe) [inline src/MongoDB.Bson/Serialization/Serializers/DictionarySerializerBase.cs:608]: I'm not sure about this, the `CreateAccumulator` method is actually using this one  in its body.

**Comment by papafe** [comment]
@rstam I have removed everything with the `Obsolete` attribute from almost everywhere in the BSON package, apart from a couple of places where I'm unsure. I've left a TODO and a comment in those places.

File: src/MongoDB.Bson/IO/JsonOutputMode.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Bson/IO/JsonOutputMode.cs:28]: Good question! Not sure either.

File: src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs:36]: We could, but... if we do anyone who has really old data might not be able to read it anymore.

I'd ask Boris and Rishit.

File: src/MongoDB.Bson/ObjectModel/BsonDocument.cs [hunk 5]
```diff
@@ -159,6 +123,7 @@ public BsonDocument(IEnumerable<BsonElement> elements)
             AddRange(elements);
         }
 
+        //TODO Are we sure we want to remove this? Having a constructor with params is useful
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonDocument.cs:126]: Agreed.

File: src/MongoDB.Bson/Serialization/Serializers/DictionarySerializerBase.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/DictionarySerializerBase.cs:608]: I think we do want to remove this method and change `CreateAccumulator` to throw `NotImplementedException`.

File: src/MongoDB.Bson/IO/JsonOutputMode.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Bson/IO/JsonOutputMode.cs:28]: FYI `Strict` is a format the C# driver used BEFORE there was any extended JSON standard, but it is very similar to `CanonicalExtendedJson`.

**Review by rstam — CHANGES_REQUESTED** [review]
Overall looks great!

File: src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs (hunk 1 above)
Comment (papafe) [reply src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs:36]: @BorisDog @rishitb-mongodb What do you think, should we remove this?

File: src/MongoDB.Bson/IO/JsonOutputMode.cs (hunk 3 above)
Comment (papafe) [reply src/MongoDB.Bson/IO/JsonOutputMode.cs:28]: @BorisDog @rishitb-mongodb Also here, do you think we should remove `JsonOutputMode.Strict`?

File: src/MongoDB.Bson/ObjectModel/BsonDocument.cs (hunk 2 above)
Comment (BorisDog) [reply src/MongoDB.Bson/ObjectModel/BsonDocument.cs:132]: Either way it fine IMHO.

File: src/MongoDB.Bson/ObjectModel/BsonDocument.cs (hunk 5 above)
Comment (papafe) [reply src/MongoDB.Bson/ObjectModel/BsonDocument.cs]: Removed the comment and left the constructor

File: src/MongoDB.Bson/Serialization/Serializers/DictionarySerializerBase.cs (hunk 4 above)
Comment (papafe) [reply src/MongoDB.Bson/Serialization/Serializers/DictionarySerializerBase.cs]: Done

File: src/MongoDB.Bson/IO/JsonOutputMode.cs (hunk 3 above)
Comment (papafe) [reply src/MongoDB.Bson/IO/JsonOutputMode.cs]: We decided to remove it

File: src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs (hunk 1 above)
Comment (papafe) [reply src/MongoDB.Bson/ObjectModel/BsonBinarySubType.cs]: We decided to keep it, for compatibility reasons

**Pull request merged**